addons/binding/org.openhab.binding.sleepiq/src/main/java/org/openhab/binding/sleepiq/handler/SleepIQCloudHandler.java

```diff
@@ -0,0 +1,322 @@
+        }
+
+        logger.debug("Setting SleepIQ cloud online");
+        updateListenerManagement();
+        updateStatus(ThingStatus.ONLINE);
+    }
+
+    /**
+     * Create a new SleepIQ cloud service connection. If a connection already exists, it will be lost.
+     *
+     * @throws LoginException if there is an error while authenticating to the service
+     */
+    private synchronized void createCloudConnection() throws LoginException {
+
+        logger.debug("Reading SleepIQ cloud binding configuration");
+        SleepIQCloudConfiguration bindingConfig = getConfigAs(SleepIQCloudConfiguration.class);
+
+        logger.debug("Creating SleepIQ client");
+        Configuration cloudConfig = new Configuration().withUsername(bindingConfig.username)
+                .withPassword(bindingConfig.password).withLogging(bindingConfig.logging);
+        cloud = SleepIQ.create(cloudConfig);
+
+        logger.info("Authenticating to SleepIQ cloud service");
+        cloud.login();
+
+        logger.info("Authentication successful");
+    }
+
+    @Override
+    public synchronized void dispose() {
+
+        logger.debug("Disposing SleepIQ cloud handler");
+
+        if (pollingJob != null && !pollingJob.isCancelled()) {
+            pollingJob.cancel(true);
+            pollingJob = null;
+        }
+
+        cloud = null;
+    }
+
+    /**
+     * Get the cloud service API.
+     *
+     * @return the service object (may be <code>null</code>)
+     */
+    private synchronized SleepIQ getCloud() {
+        return cloud;
+    }
+
+    /**
+     * Start or stop a background polling job to look for bed status updates based on the whether or not there are any
+     * listeners to notify.
+     */
+    private synchronized void updateListenerManagement() {
+
+        if (cloud == null) {
+            return;
+        }
+
+        if (!bedStatusListeners.isEmpty() && (pollingJob == null || pollingJob.isCancelled())) {
+            int pollingInterval = getConfigAs(SleepIQCloudConfiguration.class).pollingInterval;
+            pollingJob = scheduler.scheduleAtFixedRate(pollingRunnable, pollingInterval, pollingInterval,
+                    TimeUnit.SECONDS);
+        } else if (bedStatusListeners.isEmpty() && pollingJob != null && !pollingJob.isCancelled()) {
+            pollingJob.cancel(true);
+            pollingJob = null;
+        }
+    }
+
+    /**
+     * Retrieve the latest status on all beds and update all registered listeners.
+     */
+    private void publishBedStatusUpdates() {
+        publishBedStatusUpdates(null);
+    }
+
+    /**
+     * Retrieve the latest status on all beds and update the given listener. If the listener is <code>null</code>, all
+     * registered listeners will be updated.
+     *
+     * @param listener the listener to update (may be <code>null</code>)
+     */
+    private void publishBedStatusUpdates(final BedStatusListener listener) {
+
+        SleepIQ cloud = getCloud();
+        if (cloud == null) {
```

Comment:
My reasoning behind the checks is that I don't know when the platform might need to dispose of the thing and what objects still have a reference to it when it is disposed. I observed other bindings making similar null checks so I made the assumption that I should protect my code. I don't want to cause problems outside my binding by throwing runtime exceptions when I could just fail gracefully.

I will defer to project best practices here. If you would rather see an NPE in these cases, I will remove the null checks. I'm open to other suggestions as well.